### DISCUSSION TITLE
Reading array type attribute

### DISCUSSION THREAD

**Discussion by katnellayappan** [BlueBrain/HighFive#632] [Questions]
Hi,

I have a HDF5 database and one of the datasets in it has an array type attribute. I'm trying to figure out how many elements are in the array but I'm having difficulty in figuring that out.

I tried using the following code snippet but it doesn't give me the correct answer:

```cpp
HighFive::File wdb(fileName);
if (wdb.isValid()) {
      if (wdb.exist(dataSetPath)) {
        auto const& dataSet(wdb.getDataSet(dataSetPath));
        if (dataSet.isValid()) {
            auto const& attr(dataSet.getAttribute(attributeName));
            auto const& numAttrElems(attr.getSpace().getElementCount());
       }
}
```

numAttrElems is always 1 and in my database, I know that the attribute has 12 elements in the array. 

Could someone please let me know correct way to get number of elements in the array type attribute? 

-Kat

**Comment by 1uc** [comment]
HDF5 allows for something they call composite datatypes. If I understand correctly, those would allow people to store a single element which has datatype `T[12]` (i.e. a fixed size array).

Can you please try and add a bit more information about the actual HDF5 file? One way would be to use `h5dump -Hp` (which will only show information about the structure of the file). You can get it to only print information about the attribute in question by doing:
```bash
h5dump -Hp -a /path/to/attribute file.h5
```

**Reply by katnellayappan** [reply]
Thanks for the follow up. Here is the output of running `h5dump`:
```
HDF5 "input.emirtap.emir0.wdb" {
ATTRIBUTE "resIdDbgMapping" {
   DATATYPE  H5T_ARRAY { [12] H5T_STD_I32LE }
   DATASPACE  SCALAR
}
}
```

**Reply by 1uc** [reply]
Okay, this is a scalar dataspace, i.e. the number of elements is one. However, the type of that single element is an array of 12 32-bit integers. You can then get the datatype, by doing:
```
const auto& datatype = attribute.getDataType();
if(datatype.getClass() == DataTypeClass::Compound) {
  auto size_in_bytes = datatype.getSize();

  // Try to figure out how many elements this results in.
}
```
How to proceed depends on your application. With some luck you know that it's an array of 32-bit integers. If that's the case you can figure out the number of elements by dividing the size of the datatype by `sizeof(int)` (or `sizeof(std::int32_t)`).

If not, it seem you'll need to use HDF5 directly. You can find details in the HDF5 User Guide: 6.5.1.2. Properties of Composite Datatypes. Available here:
https://portal.hdfgroup.org/display/HDF5/HDF5+User%27s+Guide

You can obtain the HDF5 identifier of any object by its member `.getId()`, e.g. `datatype.getId()`.

**Comment by katnellayappan** [comment]
Thank you! That helped.

-Kat.
